test/unit/lib/plugins/aws/customResources/generateZip.test.js

```diff
@@ -0,0 +1,279 @@
+'use strict';
+
+const proxyquire = require('proxyquire').noCallThru();
+const fse = require('fs-extra');
+const sinon = require('sinon');
+const chai = require('chai');
+const fs = require('fs');
+const { getTmpDirPath, listZipFiles } = require('../../../../../utils/fs');
+const { join } = require('path');
+const { expect } = require('chai');
+const fsp = require('fs').promises;
+const sinonChai = require('sinon-chai');
+
+chai.use(sinonChai);
+
+// Stub out functions that generateZip calls to force error conditions
+const ensureArtifactStub = sinon.stub();
+const getTmpDirPathStub = sinon.stub();
+const renameStub = sinon.stub(fsp, 'rename');
+const safeMoveFile = proxyquire('../../../../../../lib/utils/fs/safeMoveFile', {
+  fs: {
+    promises: fsp,
+  },
+});
+const generateZip = proxyquire('../../../../../../lib/plugins/aws/customResources/generateZip.js', {
+  '../../../utils/ensureArtifact': ensureArtifactStub,
+  '../../../utils/fs/getTmpDirPath': getTmpDirPathStub,
+  '../../../utils/fs/safeMoveFile': safeMoveFile,
+});
+
+/**
+ * Returns and Error object that resembles the EXDEV errors that are thrown
+ * when you try to rename across device boundaries.
+ *
+ * @returns an Error object that resembles the EXDEV errors are thrown
+ */
+const createFakeExDevError = () => {
+  // Properties captured by using fs.renameSync in the Node v12.20.1 REPL
+  const fakeCrossDeviceError = new Error(
+    "Error: EXDEV: cross-device link not permitted, rename '/foo/bar' -> '/bar/baz'"
+  );
+  // This is the important property that safeMoveFile looks for
+  // to fallback to copy then rename behaviour
+  fakeCrossDeviceError.code = 'EXDEV';
+
+  // These other properties aren't used but it doesn't hurt to have an accurate error object
+  fakeCrossDeviceError.errno = -18;
+  fakeCrossDeviceError.syscall = 'rename';
+  fakeCrossDeviceError.path = '/foo/bar';
+  fakeCrossDeviceError.dest = '/bar/baz';
+  return fakeCrossDeviceError;
+};
+
+/**
+ * The name of the artefact that generateZip generates
+ */
+const artefactName = 'resource-lambda.zip';
+
+// Generate temporary directories once per test run
+// They will be cleaned up when the tests are done
+
+/**
+ * The path to the mock cache directory
+ */
+const cachePath = getTmpDirPath();
+/**
+ * The path to the directory when the zip file is generated in
+ * before it is copied to the cache path
+ */
+const sourcePath = getTmpDirPath();
+
+/**
+ * The full path to the generated artefact after is has been
+ * copied to the cache directory
+ */
+const cachedFile = join(cachePath, artefactName);
+
+describe('#generateZip()', () => {
```

Comment:
I think this should be the tests for `saveMoveFile` util specifically, can we convert it to it.

Reply:
Could I just extract the safeMoveFile parts to a separate test and simplify this test? Or are you thinking we get rid of this test altogether?

Reply:
I ended up simplifying the generateZip tests greatly and extracted the safeMoveFile tests into their own test file

Reply:
If we're about to create a test for this file, then I believe it should just confirm, that after resolution of `generateZip()`, at returned path we have an archive of which content reflects content of `resources` folder.

Confirming on other things, I believe is out of scope here, as it's handled by used utils (and not inline implementation) which have that tested in scope of their own unit tests.

Also if we follow that, we do not need to introduce any modifications to `generateZip` module